src/Shared/ServerInfrastructure/HttpCharacters.cs

```diff
@@ -1,224 +1,54 @@
-    private static readonly bool[] _token = InitializeToken();
-    private static readonly bool[] _host = InitializeHost();
-    private static readonly bool[] _fieldValue = InitializeFieldValue();
+    // ALPHA and DIGIT https://tools.ietf.org/html/rfc5234#appendix-B.1
+    private const string AlphaNumeric = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz";
 
-    internal static void Initialize()
-    {
-        // Access _alphaNumeric to initialize static fields
-        var _ = _alphaNumeric;
-    }
+    // Authority https://tools.ietf.org/html/rfc3986#section-3.2
+    // Examples:
+    // microsoft.com
+    // hostname:8080
+    // [::]:8080
+    // [fe80::]
+    // 127.0.0.1
+    // user@host.com
+    // user:password@host.com
+    private static readonly IndexOfAnyValues<byte> _allowedAuthorityBytes = IndexOfAnyValues.Create(":.-[]@0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz"u8);
 
-    private static bool[] InitializeAlphaNumeric()
-    {
-        // ALPHA and DIGIT https://tools.ietf.org/html/rfc5234#appendix-B.1
-        var alphaNumeric = new bool[_tableSize];
-        for (var c = '0'; c <= '9'; c++)
-        {
-            alphaNumeric[c] = true;
-        }
-        for (var c = 'A'; c <= 'Z'; c++)
-        {
-            alphaNumeric[c] = true;
-        }
-        for (var c = 'a'; c <= 'z'; c++)
-        {
-            alphaNumeric[c] = true;
-        }
-        return alphaNumeric;
-    }
+    // Matches Http.Sys
+    // Matches RFC 3986 except "*" / "+" / "," / ";" / "=" and "%" HEXDIG HEXDIG which are not allowed by Http.Sys
+    private static readonly IndexOfAnyValues<char> _allowedHostChars = IndexOfAnyValues.Create("!$&'()-._~" + AlphaNumeric);
 
-    private static bool[] InitializeAuthority()
-    {
-        // Authority https://tools.ietf.org/html/rfc3986#section-3.2
-        // Examples:
-        // microsoft.com
-        // hostname:8080
-        // [::]:8080
-        // [fe80::]
-        // 127.0.0.1
-        // user@host.com
-        // user:password@host.com
-        var authority = new bool[_tableSize];
-        Array.Copy(_alphaNumeric, authority, _tableSize);
-        authority[':'] = true;
-        authority['.'] = true;
-        authority['-'] = true;
-        authority['['] = true;
-        authority[']'] = true;
-        authority['@'] = true;
-        return authority;
-    }
+    // tchar https://tools.ietf.org/html/rfc7230#appendix-B
+    private static readonly IndexOfAnyValues<char> _allowedTokenChars = IndexOfAnyValues.Create("!#$%&'*+-.^_`|~" + AlphaNumeric);
+    private static readonly IndexOfAnyValues<byte> _allowedTokenBytes = IndexOfAnyValues.Create("!#$%&'*+-.^_`|~0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz"u8);
 
-    private static bool[] InitializeToken()
-    {
-        // tchar https://tools.ietf.org/html/rfc7230#appendix-B
-        var token = new bool[_tableSize];
-        Array.Copy(_alphaNumeric, token, _tableSize);
```

Comment:
Moot after the change, but it seems like this type should have had a static constructor to ensure initialization order.

Reply:
This was done via `Initialize`-method (L17 in old code) which got called from Kestrel startup, so that "relevant order" of initialization was given.

Reply:
Huh.  Can't say I've seen that pattern before.  Thanks!